{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " ![CellphoneDB Logo](https://www.cellphonedb.org/images/cellphonedb_logo_33.png) | CellphoneDB is a publicly available repository of curated receptors, ligands and their interactions. ||\n",
    " :------------- | :------------- | :-------------\n",
    "\n",
    "CellphoneDB includes subunit architecture for both ligands and receptors, representing heteromeric complexes accurately. This is crucial, as cell-cell communication relies on multi-subunit protein complexes that go beyond the binary representation used in most databases and studies.\n",
    "\n",
    "CellPhoneDB integrates existing datasets that pertain to cellular communication and new manually reviewed information. CellPhoneDB utilises information from the following data bases: [UniProt](https://www.uniprot.org/), [Ensembl](https://www.ensembl.org/), [PDB](https://www.ebi.ac.uk/pdbe/), [the IMEx consortium](https://www.imexconsortium.org/) and [IUPHAR](https://www.guidetopharmacology.org/).\n",
    "\n",
    "CellPhoneDB can be used to search for a particular ligand/receptor or interrogate your own single-cell transcriptomics data.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Install CellphoneDB package"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "pip install --force-reinstall \"git+https://github.com/ventolab/CellphoneDB.git\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## List CellphoneDB data releases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.display import HTML, display\n",
    "from cellphonedb.utils import db_releases_utils\n",
    "display(HTML(db_releases_utils.get_remote_database_versions_html()['db_releases_html_table']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Set CellphoneDB version and local directories for the database and user data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "# The default version of CellphoneDB data is the latest one, but you can change it to a previous version \n",
    "# at any point in this notebook (by re-setting the value of cpdb_version variable). \n",
    "# Please note that the format of the database from version v4.1.0 is incompatible with that of previous \n",
    "# versions, hence the lowest version number you may choose in this notebook is v4.1.0\n",
    "cpdb_version = \"v4.1.0\"\n",
    "# N.B. At the very least, please replace <your_user_id> with your user id\n",
    "cpdb_dir = os.path.join(\"/Users/rp23/.cpdb/releases\", cpdb_version)\n",
    "# If you generated your own CellphoneDB database file, please replace the default path below to the your file's path\n",
    "cpdb_file_path = os.path.join(cpdb_dir, \"cellphonedb.zip\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Download CellphoneDB database from https://github.com/ventolab/cellphonedb-data/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from cellphonedb.utils import db_utils\n",
    "db_utils.download_database(cpdb_dir, cpdb_version)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Search CellphoneDB Interactions\n",
    "#### Search CellphoneDB interactions by (a comma- or space-separated list of): \n",
    "* Ensembl ID (e.g. ENSG00000165029), \n",
    "* Gene name (e.g. ABCA1), \n",
    "* UniProt ID (e.g. KLRG2_HUMAN), \n",
    "* UniProt Accession (e.g. A4D1S0) or \n",
    "* Complex name (e.g. 12oxoLeukotrieneB4_byPTGR1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from cellphonedb.utils import file_utils, search_utils\n",
    "from IPython.display import HTML, display\n",
    "# Search CellphoneDB interactions by (a comma- or space-separated list of):\n",
    "# Ensembl ID (e.g. ENSG00000165029), Gene name (e.g. ABCA1), UniProt ID (e.g. KLRG2_HUMAN), \n",
    "# UniProt Accession (e.g. A4D1S0) or Complex name (e.g. 12oxoLeukotrieneB4_byPTGR1)\n",
    "(results, complex_name2proteins_text) = \\\n",
    "    search_utils.search(query_str = 'D17S1718,ENSG00000134780,integrin_a10b1_complex', \n",
    "                        cpdb_file_path = cpdb_file_path)\n",
    "# Display results in a html table\n",
    "# Note: Mouse over complex names to see constituent proteins\n",
    "display(HTML(search_utils.get_html_table(results, complex_name2proteins_text)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run Basic Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id_cp_interaction</th>\n",
       "      <th>partner_a</th>\n",
       "      <th>partner_b</th>\n",
       "      <th>Lymphoid|SOX9_prolif</th>\n",
       "      <th>SOX9_prolif|Lymphoid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>952</th>\n",
       "      <td>CPI-SS05632AB58</td>\n",
       "      <td>simple:P13591</td>\n",
       "      <td>simple:P11362</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1095</th>\n",
       "      <td>CPI-SC02DEB4893</td>\n",
       "      <td>simple:P01579</td>\n",
       "      <td>complex:Type_II_IFNR</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1658</th>\n",
       "      <td>CPI-SS0D168412F</td>\n",
       "      <td>simple:P48023</td>\n",
       "      <td>simple:P25445</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1165</th>\n",
       "      <td>CPI-SS0E292C126</td>\n",
       "      <td>simple:Q12918</td>\n",
       "      <td>simple:Q9UHP7</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1651</th>\n",
       "      <td>CPI-SS02C509183</td>\n",
       "      <td>simple:Q06643</td>\n",
       "      <td>simple:P36941</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>970</th>\n",
       "      <td>CPI-SC0E029AB31</td>\n",
       "      <td>simple:Q03405</td>\n",
       "      <td>complex:integrin_a4b1_complex</td>\n",
       "      <td>0.0</td>\n",
       "      <td>22.137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>971</th>\n",
       "      <td>CPI-SC0B2557C9D</td>\n",
       "      <td>simple:P19320</td>\n",
       "      <td>complex:integrin_a4b1_complex</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>972</th>\n",
       "      <td>CPI-SC034EE8AAA</td>\n",
       "      <td>simple:P02751</td>\n",
       "      <td>complex:integrin_a4b7_complex</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>973</th>\n",
       "      <td>CPI-SC056FC1F11</td>\n",
       "      <td>simple:Q13477</td>\n",
       "      <td>complex:integrin_a4b7_complex</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2922</th>\n",
       "      <td>CPI-CS0FC56FE03</td>\n",
       "      <td>complex:b-PEA_byDDC</td>\n",
       "      <td>simple:Q96RJ0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2434 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     id_cp_interaction            partner_a                      partner_b  \\\n",
       "952    CPI-SS05632AB58        simple:P13591                  simple:P11362   \n",
       "1095   CPI-SC02DEB4893        simple:P01579           complex:Type_II_IFNR   \n",
       "1658   CPI-SS0D168412F        simple:P48023                  simple:P25445   \n",
       "1165   CPI-SS0E292C126        simple:Q12918                  simple:Q9UHP7   \n",
       "1651   CPI-SS02C509183        simple:Q06643                  simple:P36941   \n",
       "...                ...                  ...                            ...   \n",
       "970    CPI-SC0E029AB31        simple:Q03405  complex:integrin_a4b1_complex   \n",
       "971    CPI-SC0B2557C9D        simple:P19320  complex:integrin_a4b1_complex   \n",
       "972    CPI-SC034EE8AAA        simple:P02751  complex:integrin_a4b7_complex   \n",
       "973    CPI-SC056FC1F11        simple:Q13477  complex:integrin_a4b7_complex   \n",
       "2922   CPI-CS0FC56FE03  complex:b-PEA_byDDC                  simple:Q96RJ0   \n",
       "\n",
       "      Lymphoid|SOX9_prolif  SOX9_prolif|Lymphoid  \n",
       "952                  100.0                 0.000  \n",
       "1095                 100.0                 0.000  \n",
       "1658                 100.0                 0.000  \n",
       "1165                 100.0                 0.000  \n",
       "1651                 100.0                 0.000  \n",
       "...                    ...                   ...  \n",
       "970                    0.0                22.137  \n",
       "971                    0.0                 0.000  \n",
       "972                    0.0                 0.000  \n",
       "973                    0.0                 0.000  \n",
       "2922                   0.0                 0.000  \n",
       "\n",
       "[2434 rows x 5 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Please populate the following variables before executing the analysis\n",
    "meta_file_path = None\n",
    "counts_file_path = None\n",
    "output_path = None\n",
    "output_path = \"/Users/rp23/.cpdb/user_files/out\"\n",
    "counts_file_path=\"/Users/rp23/.cpdb/tests/data/examples/endometrium_v1/counts.h5ad\"\n",
    "meta_file_path=\"/Users/rp23/.cpdb/tests/data/examples/endometrium_v1/meta.tsv\"\n",
    "degs_file_path=\"/Users/rp23/.cpdb/tests/data/examples/endometrium_v1/degs_in_epithelials.tsv\"\n",
    "microenvs_file_path = \"/Users/rp23/.cpdb/tests/data/examples/endometrium_v1/microenviroments.tsv\"\n",
    "cpdb_file_path = os.path.join(cpdb_dir, \"cellphonedb.zip\")\n",
    "# Execute basic analysis\n",
    "from cellphonedb.src.core.methods import cpdb_analysis_method\n",
    "means, deconvoluted, deconvoluted_percents, interaction_scores = cpdb_analysis_method.call(\n",
    "    cpdb_file_path = cpdb_file_path, \n",
    "    meta_file_path = meta_file_path, \n",
    "    counts_file_path = counts_file_path,\n",
    "    counts_data = 'hgnc_symbol',\n",
    "    output_path = output_path,\n",
    "    microenvs_file_path = None,\n",
    "    separator = \"|\",\n",
    "    threshold = 0.1,\n",
    "    result_precision = 3,\n",
    "    debug = False,\n",
    "    output_suffix = None,\n",
    "    score_interactions = True,\n",
    "    threads = 4)\n",
    "# print(means.info)\n",
    "# print(deconvoluted.info)\n",
    "# print(deconvoluted_percents.info)\n",
    "# print(interaction_scores.info)\n",
    "example_table = interaction_scores[['id_cp_interaction','partner_a','partner_b','Lymphoid|SOX9_prolif', \n",
    "                    'SOX9_prolif|Lymphoid']].sort_values('Lymphoid|SOX9_prolif', ascending = False)\n",
    "example_table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run Statistical Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading user files...\n",
      "The following user files were loaded successfully:\n",
      "/Users/rp23/.cpdb/tests/data/examples/endometrium_v1/counts.h5ad\n",
      "/Users/rp23/.cpdb/tests/data/examples/endometrium_v1/meta.tsv\n",
      "[ ][CORE][03/05/23-10:39:11][INFO] [Cluster Statistical Analysis] Threshold:0.1 Iterations:1000 Debug-seed:-1 Threads:4 Precision:3\n",
      "[ ][CORE][03/05/23-10:39:18][INFO] Running Real Analysis\n",
      "[ ][CORE][03/05/23-10:39:18][INFO] Running Statistical Analysis\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  2%|██▊                                                                                                                    | 24/1000 [00:14<07:41,  2.11it/s]"
     ]
    }
   ],
   "source": [
    "# Please populate the following variables before executing the analysis\n",
    "meta_file_path = None\n",
    "counts_file_path = None\n",
    "output_path = None\n",
    "output_path = \"/Users/rp23/.cpdb/user_files/out\"\n",
    "counts_file_path=\"/Users/rp23/.cpdb/tests/data/examples/endometrium_v1/counts.h5ad\"\n",
    "meta_file_path=\"/Users/rp23/.cpdb/tests/data/examples/endometrium_v1/meta.tsv\"\n",
    "degs_file_path=\"/Users/rp23/.cpdb/tests/data/examples/endometrium_v1/degs_in_epithelials.tsv\"\n",
    "microenvs_file_path = \"/Users/rp23/.cpdb/tests/data/examples/endometrium_v1/microenviroments.tsv\"\n",
    "cpdb_file_path = os.path.join(cpdb_dir, \"cellphonedb.zip\")\n",
    "# Execute statistical analysis\n",
    "from cellphonedb.src.core.methods import cpdb_statistical_analysis_method\n",
    "deconvoluted, deconvoluted_percents, means, pvalues, significant_means, interaction_scores = \\\n",
    "    cpdb_statistical_analysis_method.call(\n",
    "        cpdb_file_path = cpdb_file_path, \n",
    "        meta_file_path = meta_file_path, \n",
    "        counts_file_path = counts_file_path,\n",
    "        counts_data = 'hgnc_symbol',\n",
    "        output_path = output_path,\n",
    "        microenvs_file_path = None,\n",
    "        iterations = 1000,\n",
    "        threshold = 0.1,\n",
    "        threads = 4,\n",
    "        debug_seed = -1,\n",
    "        result_precision = 3,\n",
    "        pvalue = 0.05,\n",
    "        subsampling = False,\n",
    "        subsampling_log = False,\n",
    "        subsampling_num_pc = 100,\n",
    "        subsampling_num_cells = None,\n",
    "        separator = '|',\n",
    "        debug = False,\n",
    "        output_suffix = None,\n",
    "        score_interactions = True)\n",
    "# print(deconvoluted.info)\n",
    "# print(deconvoluted_percents.info)\n",
    "# print(means.info)\n",
    "# print(pvalues.info)\n",
    "# print(significant_means.info)\n",
    "# print(interaction_scores.info)\n",
    "example_table = interaction_scores[['id_cp_interaction','partner_a','partner_b','Lymphoid|SOX9_prolif', \n",
    "                    'SOX9_prolif|Lymphoid']].sort_values('Lymphoid|SOX9_prolif', ascending = False)\n",
    "example_table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run Differential Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[ ][CORE][03/05/23-10:37:11][INFO] [Cluster DEGs Analysis] Threshold:0.1 Precision:3\n",
      "Reading user files...\n",
      "WARNING: DEGs expects 2 columns and got 13. Dropping extra columns.\n",
      "The following user files were loaded successfully:\n",
      "/Users/rp23/.cpdb/tests/data/examples/endometrium_v1/counts.h5ad\n",
      "/Users/rp23/.cpdb/tests/data/examples/endometrium_v1/meta.tsv\n",
      "/Users/rp23/.cpdb/tests/data/examples/endometrium_v1/degs_in_epithelials.tsv\n",
      "[ ][CORE][03/05/23-10:37:33][INFO] Running Real Analysis\n",
      "[ ][CORE][03/05/23-10:37:33][INFO] Running DEGs-based Analysis\n",
      "[ ][CORE][03/05/23-10:37:34][INFO] Building results\n",
      "[ ][CORE][03/05/23-10:37:35][INFO] Scoring interactions: Filtering genes per cell type..\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 14/14 [00:00<00:00, 20.89it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[ ][CORE][03/05/23-10:37:35][INFO] Scoring interactions: Calculating mean expression of each gene per group/cell type..\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 14/14 [00:00<00:00, 93.00it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[ ][CORE][03/05/23-10:37:36][INFO] Scoring interactions: Calculating scores for all interactions and cell types..\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 196/196 [00:40<00:00,  4.90it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved deconvoluted_result to /Users/rp23/.cpdb/user_files/out/degs_analysis_deconvoluted_result_05_03_2023_103817.txt\n",
      "Saved deconvoluted_percents to /Users/rp23/.cpdb/user_files/out/degs_analysis_deconvoluted_percents_05_03_2023_103817.txt\n",
      "Saved means_result to /Users/rp23/.cpdb/user_files/out/degs_analysis_means_result_05_03_2023_103817.txt\n",
      "Saved relevant_interactions_result to /Users/rp23/.cpdb/user_files/out/degs_analysis_relevant_interactions_result_05_03_2023_103817.txt\n",
      "Saved significant_means to /Users/rp23/.cpdb/user_files/out/degs_analysis_significant_means_05_03_2023_103817.txt\n",
      "Saved interaction_scores to /Users/rp23/.cpdb/user_files/out/degs_analysis_interaction_scores_05_03_2023_103817.txt\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id_cp_interaction</th>\n",
       "      <th>partner_a</th>\n",
       "      <th>partner_b</th>\n",
       "      <th>Lymphoid|SOX9_prolif</th>\n",
       "      <th>SOX9_prolif|Lymphoid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>952</th>\n",
       "      <td>CPI-SS05632AB58</td>\n",
       "      <td>simple:P13591</td>\n",
       "      <td>simple:P11362</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1095</th>\n",
       "      <td>CPI-SC02DEB4893</td>\n",
       "      <td>simple:P01579</td>\n",
       "      <td>complex:Type_II_IFNR</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1658</th>\n",
       "      <td>CPI-SS0D168412F</td>\n",
       "      <td>simple:P48023</td>\n",
       "      <td>simple:P25445</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1165</th>\n",
       "      <td>CPI-SS0E292C126</td>\n",
       "      <td>simple:Q12918</td>\n",
       "      <td>simple:Q9UHP7</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1651</th>\n",
       "      <td>CPI-SS02C509183</td>\n",
       "      <td>simple:Q06643</td>\n",
       "      <td>simple:P36941</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>970</th>\n",
       "      <td>CPI-SC0E029AB31</td>\n",
       "      <td>simple:Q03405</td>\n",
       "      <td>complex:integrin_a4b1_complex</td>\n",
       "      <td>0.0</td>\n",
       "      <td>22.137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>971</th>\n",
       "      <td>CPI-SC0B2557C9D</td>\n",
       "      <td>simple:P19320</td>\n",
       "      <td>complex:integrin_a4b1_complex</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>972</th>\n",
       "      <td>CPI-SC034EE8AAA</td>\n",
       "      <td>simple:P02751</td>\n",
       "      <td>complex:integrin_a4b7_complex</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>973</th>\n",
       "      <td>CPI-SC056FC1F11</td>\n",
       "      <td>simple:Q13477</td>\n",
       "      <td>complex:integrin_a4b7_complex</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2922</th>\n",
       "      <td>CPI-CS0FC56FE03</td>\n",
       "      <td>complex:b-PEA_byDDC</td>\n",
       "      <td>simple:Q96RJ0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2434 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     id_cp_interaction            partner_a                      partner_b  \\\n",
       "952    CPI-SS05632AB58        simple:P13591                  simple:P11362   \n",
       "1095   CPI-SC02DEB4893        simple:P01579           complex:Type_II_IFNR   \n",
       "1658   CPI-SS0D168412F        simple:P48023                  simple:P25445   \n",
       "1165   CPI-SS0E292C126        simple:Q12918                  simple:Q9UHP7   \n",
       "1651   CPI-SS02C509183        simple:Q06643                  simple:P36941   \n",
       "...                ...                  ...                            ...   \n",
       "970    CPI-SC0E029AB31        simple:Q03405  complex:integrin_a4b1_complex   \n",
       "971    CPI-SC0B2557C9D        simple:P19320  complex:integrin_a4b1_complex   \n",
       "972    CPI-SC034EE8AAA        simple:P02751  complex:integrin_a4b7_complex   \n",
       "973    CPI-SC056FC1F11        simple:Q13477  complex:integrin_a4b7_complex   \n",
       "2922   CPI-CS0FC56FE03  complex:b-PEA_byDDC                  simple:Q96RJ0   \n",
       "\n",
       "      Lymphoid|SOX9_prolif  SOX9_prolif|Lymphoid  \n",
       "952                  100.0                 0.000  \n",
       "1095                 100.0                 0.000  \n",
       "1658                 100.0                 0.000  \n",
       "1165                 100.0                 0.000  \n",
       "1651                 100.0                 0.000  \n",
       "...                    ...                   ...  \n",
       "970                    0.0                22.137  \n",
       "971                    0.0                 0.000  \n",
       "972                    0.0                 0.000  \n",
       "973                    0.0                 0.000  \n",
       "2922                   0.0                 0.000  \n",
       "\n",
       "[2434 rows x 5 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Please populate the following variables before executing the analysis\n",
    "meta_file_path = None\n",
    "counts_file_path = None\n",
    "degs_file_path = None\n",
    "output_path = None\n",
    "output_path = \"/Users/rp23/.cpdb/user_files/out\"\n",
    "counts_file_path=\"/Users/rp23/.cpdb/tests/data/examples/endometrium_v1/counts.h5ad\"\n",
    "meta_file_path=\"/Users/rp23/.cpdb/tests/data/examples/endometrium_v1/meta.tsv\"\n",
    "degs_file_path=\"/Users/rp23/.cpdb/tests/data/examples/endometrium_v1/degs_in_epithelials.tsv\"\n",
    "microenvs_file_path = \"/Users/rp23/.cpdb/tests/data/examples/endometrium_v1/microenviroments.tsv\"\n",
    "cpdb_file_path = os.path.join(cpdb_dir, \"cellphonedb.zip\")\n",
    "# Execute differential analysis\n",
    "from cellphonedb.src.core.methods import cpdb_degs_analysis_method\n",
    "deconvoluted, deconvoluted_percents, means, relevant_interactions, significant_means, interaction_scores = \\\n",
    "    cpdb_degs_analysis_method.call(\n",
    "        cpdb_file_path = cpdb_file_path, \n",
    "        meta_file_path = meta_file_path, \n",
    "        counts_file_path = counts_file_path,\n",
    "        degs_file_path = degs_file_path,\n",
    "        counts_data = 'hgnc_symbol',\n",
    "        microenvs_file_path=None,\n",
    "        threshold = 0.1,\n",
    "        result_precision = 3,\n",
    "        separator = '|',\n",
    "        debug = False,\n",
    "        output_path = output_path,\n",
    "        output_suffix = None,\n",
    "        score_interactions = True,\n",
    "        threads = 4)\n",
    "# print(deconvoluted.info)\n",
    "# print(means.info)\n",
    "# print(relevant_interactions.info)\n",
    "# print(significant_means.info)\n",
    "# print(interaction_scores.info)\n",
    "example_table = interaction_scores[['id_cp_interaction','partner_a','partner_b','Lymphoid|SOX9_prolif', \n",
    "                    'SOX9_prolif|Lymphoid']].sort_values('Lymphoid|SOX9_prolif', ascending = False)\n",
    "example_table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plot Statistical Analysis results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "from ktplotspy.plot import plot_cpdb, plot_cpdb_heatmap\n",
    "from cellphonedb.utils import file_utils\n",
    "import os\n",
    "\n",
    "meta_fp='test_meta.txt'\n",
    "# counts_fn='test_counts.txt'\n",
    "counts_fp='test.h5ad'\n",
    "# Create AnnData object with obs set to a DataFrame containing data from meta_fn\n",
    "adata = file_utils.get_counts_meta_adata(counts_fp, meta_fp)\n",
    "\n",
    "# Exmaple dot plot\n",
    "g1 = plot_cpdb(\n",
    "        adata=adata,\n",
    "        cell_type1=\"Myeloid\",\n",
    "        # '.' means any cell type\n",
    "        cell_type2=\".\",\n",
    "        means=means,\n",
    "        pvals=pvalues,\n",
    "        celltype_key=\"cell_type\",\n",
    "        genes=[\"FN1\", \"integrin-a5b1-complex\",\"COLEC12\"],\n",
    "        title=\"Example dot plot\"\n",
    "    )\n",
    "\n",
    "# Example heatmap\n",
    "g2 = plot_cpdb_heatmap(\n",
    "        adata=adata,\n",
    "        pvals=pvalues,\n",
    "        celltype_key=\"cell_type\",\n",
    "        log1p_transform=True,\n",
    "        title=\"Example heatmap\"\n",
    "    )\n",
    "g1, g2\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Search for interactions in Statistical or Differential Analysis Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Search results of either statistical or DEG analysis for relevant interactions matching any of:\n",
    "        1. A gene in genes\n",
    "        2. A complex containing a gene in genes\n",
    "        3. An interaction name in interactions (e.g. 12oxoLeukotrieneB4_byPTGR1)\n",
    "    where at least one pair of cell types containing one cell type from cell_types_1\n",
    "    and one cell type from cell_types_2 has a significant mean.\n",
    "    NB. If genes and interactions are empty, and cell_types_1 and cell_types_2 are both set to \"All\"\n",
    "    then all relevant interactions are returned.\n",
    "\"\"\"\n",
    "from IPython.display import HTML, display\n",
    "from cellphonedb.utils import search_utils\n",
    "\n",
    "cell_types_1=\"All\"\n",
    "cell_types_2=\"All\"\n",
    "genes=None\n",
    "interactions=None\n",
    "\n",
    "cell_types_1=['Fibroblast dS','EVT_2']\n",
    "cell_types_2=['epi_Ciliated','dS2']\n",
    "genes=['DKK1']\n",
    "interactions=['DKK1_LRP6','LPAR2_ADGRE5','TGFB1_TGFbeta_receptor2']\n",
    "separator=\"|\"\n",
    "# Set long_format to True to transpose the results table (so that cell type pairs are shown in a single column)\n",
    "long_format = False\n",
    "\n",
    "search_results = search_utils.search_analysis_results(\n",
    "        query_cell_types_1=cell_types_1,\n",
    "        query_cell_types_2=cell_types_2,\n",
    "        query_genes=genes,\n",
    "        query_interactions=interactions,\n",
    "        significant_means=significant_means,\n",
    "        deconvoluted=deconvoluted,\n",
    "        separator=separator,\n",
    "        long_format=long_format\n",
    ")\n",
    "display(HTML(search_results.to_html(index=False)))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  },
  "vscode": {
   "interpreter": {
    "hash": "d4be38938b4384b4b72f771d57e174ff27b15af44a35842c7bbd2be6ebd16de6"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
